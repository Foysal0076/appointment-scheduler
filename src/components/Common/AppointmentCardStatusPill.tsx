import clsx from 'clsx'

import { AppointmentStatusType } from '@/utils/types/appointment.types'

type Props = {
  status: AppointmentStatusType
}

const AppointmentCardStatusPill = ({ status }: Props) => {
  return (
    <div
      className={clsx(
        'select-none rounded-full border px-2 py-1 text-sm font-bold capitalize tracking-wide text-white transition-colors duration-200',
        { 'border-success-500 bg-success-400': status === 'approved' },
        { 'border-danger-500 bg-danger-400': status === 'rejected' },
        { 'border-warning-500 bg-warning-400': status === 'cancelled' },
        { 'border-info-500 bg-info-400': status === 'pending' }
      )}>
      {status}
    </div>
  )
}

export default AppointmentCardStatusPill
