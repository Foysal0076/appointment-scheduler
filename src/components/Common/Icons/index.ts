export { SearchIcon } from './SearchIcon'
export { ToggleDarkIcon } from './ToggleDarkIcon'
export { ToggleLightIcon } from './ToggleLightIcon'
