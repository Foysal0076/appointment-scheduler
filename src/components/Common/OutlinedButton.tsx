import { cva } from 'class-variance-authority'
import clsx from 'clsx'

import Spinner from '@/components/Common/Spinner'

type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  color?: 'primary' | 'secondary' | 'tertiary' | 'danger' | 'success'
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl'
  fullWidth?: boolean
  loading?: boolean
}

const ButtonVariants = cva(
  'rounded bg-none border-2 focus:ring-4 disabled:active:transform-none disabled:cursor-not-allowed dark:hover:bg-neutral-20/50 transition-all duration-300 text-nowrap focus:outline-none font-medium active:scale-[.9]',
  {
    variants: {
      color: {
        primary:
          'hover:bg-primary-50 focus:ring-primary-100 border-primary-500/80 text-primary-500  focus:ring-primary-100 dark:text-primary-200 dark:border-primary-200 disabled:text-primary-200 disabled:border-primary-100 disabled:hover:bg-transparent dark:disabled:bg-primary-50/20 dark:disabled:border-primary-50/10 dark:disabled:text-primary-100/60',
        secondary:
          'hover:bg-secondary-50 focus:ring-secondary-100 border-secondary-500/80 text-secondary-500  focus:ring-secondary-100 dark:text-secondary-200 dark:border-secondary-200 disabled:text-secondary-200 disabled:border-secondary-100 disabled:hover:bg-transparent dark:disabled:bg-secondary-50/20 dark:disabled:border-secondary-50/10 dark:disabled:text-secondary-100/60',
        tertiary:
          'hover:bg-tertiary-50 focus:ring-tertiary-100 border-tertiary-500/80 text-tertiary-500  focus:ring-tertiary-100 dark:text-tertiary-200 dark:border-tertiary-200 disabled:text-tertiary-200 disabled:border-tertiary-100 disabled:hover:bg-transparent dark:disabled:bg-tertiary-50/20 dark:disabled:border-tertiary-50/10 dark:disabled:text-tertiary-100/60',
        danger:
          'hover:bg-danger-50 focus:ring-danger-100 border-danger-500/80 text-danger-500  focus:ring-danger-100 dark:text-danger-200 dark:border-danger-200 disabled:text-danger-200 disabled:border-danger-100 disabled:hover:bg-transparent dark:disabled:bg-danger-50/20 dark:disabled:border-danger-50/10 dark:disabled:text-danger-100/60',
        success:
          'hover:bg-success-50 focus:ring-success-100 border-success-500/80 text-success-500  focus:ring-success-100 dark:text-success-200 dark:border-success-200 disabled:text-success-200 disabled:border-success-100 disabled:hover:bg-transparent dark:disabled:bg-success-50/20 dark:disabled:border-success-50/10 dark:disabled:text-success-100/60',
      },
      size: {
        xs: 'h-[1.75rem] text-xs px-4',
        sm: 'h-[2rem] text-sm px-4',
        md: 'h-[2.5rem] text-md px-4',
        lg: 'h-[3.125rem] text-lg px-4',
        xl: 'h-[3.5rem] text-lg px-4',
      },
    },
    defaultVariants: {
      color: 'primary',
      size: 'md',
    },
  }
)

const OutlinedButton = ({
  color = 'primary',
  size = 'md',
  loading = false,
  className = '',
  children,
  ...buttonProps
}: ButtonProps) => {
  return (
    <button
      className={clsx(
        `relative select-none`,
        `${ButtonVariants({ color, size })}`,
        `${className}`
      )}
      disabled={loading}
      {...buttonProps}>
      {loading && (
        <div className='absolute inset-0 flex items-center justify-center'>
          <Spinner className='text-current' />
        </div>
      )}
      <div
        className={clsx(`visible font-Rubik`, {
          invisible: loading,
          'opacity-0': loading,
        })}>
        {children}
      </div>
    </button>
  )
}

export default OutlinedButton
