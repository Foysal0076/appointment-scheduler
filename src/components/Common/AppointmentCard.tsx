import { useState } from 'react'
import { toast } from 'react-toastify'

import AppointmentCardStatusPill from '@/components/Common/AppointmentCardStatusPill'
import OutlinedButton from '@/components/Common/OutlinedButton'
import { PutAppointmentBody } from '@/redux/apiQueries/apiQueries.type'
import {
  useDeleteAppointmentMutation,
  useUpdateAppointmentMutation,
} from '@/redux/apiQueries/appointmentQueries'
import { formatDate, formatTime } from '@/utils/helpers'
import {
  AppointmentItem,
  AppointmentStatusType,
  HourFormat,
} from '@/utils/types/appointment.types'

type AppointmentCardProps = AppointmentItem & {
  userId: string
  hourFormat?: HourFormat
}

const AppointmentCard = ({
  id,
  title,
  description,
  audioMessage,
  hostInfo,
  guestInfo,
  userId,
  status,
  startTime,
  endTime,
}: AppointmentCardProps) => {
  const isHost = hostInfo.id === userId
  const isRejected = status === 'rejected'
  const isPast = new Date().getTime() > endTime

  const [updateAppointment, { isLoading: isUpdatingAppointment }] =
    useUpdateAppointmentMutation()

  const [deleteAppointment, { isLoading: isDeletingAppointment }] =
    useDeleteAppointmentMutation()

  const [loadingStatus, setLoadingStatus] =
    useState<AppointmentStatusType | null>(null)

  const onCancelAppointment = async () => {
    try {
      //check if the meeting is already started
      const currentTime = new Date().getTime()
      if (currentTime > startTime) {
        toast.error('You can not cancel a meeting that is past current time')
        return
      }
      console.log('cancel appointment')
      const body: PutAppointmentBody = {
        id,
        status: 'cancelled',
      }
      await updateAppointment(body)
    } catch (error) {
      console.log(error)
      toast.error('Something went wrong')
    }
  }

  const onAcceptOrDeclineAppointment = async (
    status: AppointmentStatusType
  ) => {
    try {
      setLoadingStatus(status)
      const body: PutAppointmentBody = {
        id,
        status,
      }
      await updateAppointment(body)
    } catch (error) {
      console.log(error)
      toast.error('Something went wrong')
    } finally {
      setLoadingStatus(null)
    }
  }

  const onDeleteAppointment = async () => {
    try {
      await deleteAppointment(id)
    } catch (error) {
      console.log(error)
      toast.error('Something went wrong')
    }
  }

  return (
    <div className='relative flex h-full flex-col justify-between gap-4 rounded-lg border border-neutral-30 bg-surface-0 p-4 shadow-md dark:bg-surface-100'>
      <div className='absolute right-2 top-2'>
        <AppointmentCardStatusPill status={status} />
      </div>
      <div className='grow'>
        <h2 className='h3 mb-2'>{formatDate(startTime)}</h2>
        <p className='h5 mb-4 font-bold'>
          {formatTime(startTime)} - {formatTime(endTime)}
        </p>
        <h3 className='h6 mb-4'>{title}</h3>
        <p className='mb-4 line-clamp-4 text-base text-neutral-400'>
          {description}
        </p>
        <div className='mb-4'>
          {audioMessage && <audio src={audioMessage} controls />}
        </div>
      </div>
      <div>
        <p className='mb-4 text-xs text-neutral-700 md:text-sm'>
          {isHost && isRejected
            ? `${guestInfo.fullname} rejected this appointment`
            : `Participants: ${hostInfo.fullname}, ${guestInfo.fullname}`}
        </p>
        <div className='flex justify-between gap-4'>
          {isHost ? (
            <>
              <OutlinedButton
                onClick={onDeleteAppointment}
                color='danger'
                size='sm'
                loading={isDeletingAppointment}>
                Delete
              </OutlinedButton>
              <OutlinedButton
                onClick={onCancelAppointment}
                color='tertiary'
                size='sm'
                disabled={status === 'cancelled' || isPast}
                loading={isUpdatingAppointment}>
                {status === 'cancelled' ? 'Cancelled' : 'Cancel'}
              </OutlinedButton>
            </>
          ) : (
            <>
              <OutlinedButton
                onClick={() => onAcceptOrDeclineAppointment('rejected')}
                color='danger'
                size='sm'
                disabled={status === 'rejected' || isPast}
                loading={loadingStatus === 'rejected'}>
                Decline
              </OutlinedButton>
              <OutlinedButton
                onClick={() => onAcceptOrDeclineAppointment('approved')}
                color='success'
                size='sm'
                disabled={status === 'approved' || isPast}
                loading={loadingStatus === 'approved'}>
                {status === 'approved' ? 'Accepted' : 'Accept'}
              </OutlinedButton>
            </>
          )}
        </div>
      </div>
    </div>
  )
}

export default AppointmentCard
