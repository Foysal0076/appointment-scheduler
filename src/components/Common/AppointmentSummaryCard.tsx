import Button from '@/components/Common/Button'
import OutlinedButton from '@/components/Common/OutlinedButton'
import { HOUR_FORMAT } from '@/utils/constants/appointment.constants'
import { formatDate, formatTime } from '@/utils/helpers'
import { HourFormat } from '@/utils/types/appointment.types'

type AppointmentSummaryCardProps = {
  title: string
  description: string
  host: string // the person who scheduled the appointment
  guest: string // the person who is invited for the appointment
  audioFile?: File | Blob | null
  startTime: Date
  endTime: Date
  hourFormat?: HourFormat
  onSubmit: () => void
  onCancel: () => void
  loading?: boolean
}

const AppointmentSummaryCard = ({
  host,
  guest,
  audioFile,
  startTime,
  endTime,
  title,
  description,
  hourFormat = HOUR_FORMAT.h12,
  onSubmit,
  onCancel,
  loading,
}: AppointmentSummaryCardProps) => {
  return (
    <div className='flex h-full flex-col justify-between rounded-lg border border-neutral-30 bg-surface-0 p-4 shadow-md dark:bg-surface-100'>
      <div className='grow'>
        <h2 className='h3 mb-2'>{formatDate(startTime)}</h2>
        <p className='h5 mb-4 inline font-bold'>
          {formatTime(startTime, hourFormat)} -{' '}
          {formatTime(endTime, hourFormat)}
        </p>
        <h3 className='h6 mb-4'>{title}</h3>
        <p className='mb-4 line-clamp-4'>{description}</p>
        <div className='mb-4'>
          {audioFile && <audio src={URL.createObjectURL(audioFile)} controls />}
        </div>
        <p className='mb-4 inline text-sm text-neutral-700'>
          Participants: {host} and {guest}
        </p>
      </div>
      <div className='flex justify-end gap-4'>
        <>
          <OutlinedButton color='danger' size='sm' onClick={onCancel}>
            Cancel
          </OutlinedButton>
          <Button size='sm' onClick={onSubmit} loading={loading}>
            Confirm
          </Button>
        </>
      </div>
    </div>
  )
}

export default AppointmentSummaryCard
