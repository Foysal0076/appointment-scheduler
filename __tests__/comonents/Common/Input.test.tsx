import { Input } from '@/components/Common/Input'
import { fireEvent, render } from '@testing-library/react'

describe('Input component', () => {
  test('renders without crashing', () => {
    const { container } = render(<Input />)
    expect(container).toBeInTheDocument()
  })

  test('applies default styles', () => {
    const { container } = render(<Input />)
    const inputElement = container.querySelector('input')
    expect(inputElement).toHaveClass('rounded-sm border-none bg-surface-50')
  })

  test('renders label when provided', () => {
    const { container } = render(<Input label='Username' />)
    const labelElement = container.querySelector('label')
    expect(labelElement).toBeInTheDocument()
    expect(labelElement).toHaveTextContent('Username')
  })

  test('applies error styles when error is provided', () => {
    const { container } = render(<Input error='Invalid input' />)
    const inputElement = container.querySelector('input')
    expect(inputElement).toHaveClass('ring-danger-500 focus:ring-danger-500')
  })

  test('calls onClick handler when start adornment button is clicked', () => {
    const handleClick = jest.fn()
    const { container } = render(
      <Input
        startAdornment={{
          adornment: <span>🔍</span>,
          onClick: handleClick,
        }}
      />
    )
    const startAdornmentButton = container.querySelector(
      'button[aria-label="start-adornment btn"]'
    )
    if (startAdornmentButton) {
      fireEvent.click(startAdornmentButton)
      expect(handleClick).toHaveBeenCalled()
    }
  })

  test('calls onClick handler when end adornment button is clicked', () => {
    const handleClick = jest.fn()
    const { container } = render(
      <Input
        endAdornment={{
          adornment: <span>👁️</span>,
          onClick: handleClick,
        }}
      />
    )
    const endAdornmentButton = container.querySelector(
      'button[aria-label="end-adornment btn"]'
    )
    if (endAdornmentButton) {
      fireEvent.click(endAdornmentButton)
      expect(handleClick).toHaveBeenCalled()
    }
  })

  test('matches snapshot', () => {
    const { asFragment } = render(<Input />)
    expect(asFragment()).toMatchSnapshot()
  })
})
