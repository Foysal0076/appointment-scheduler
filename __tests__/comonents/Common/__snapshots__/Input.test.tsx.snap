// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="custom-input group flex flex-col gap-1"
  >
    <div
      class="relative flex flex-col"
    >
      <input
        class="rounded-sm border-none bg-surface-50 px-4 text-neutral-600 outline-none ring-1 ring-surface-400 placeholder:text-neutral-100 focus:placeholder:text-primary-100 disabled:bg-neutral-30 disabled:text-neutral-100 dark:bg-surface-200 dark:disabled:bg-surface-400  ring-neutral-50 focus:ring-primary-500 dark:focus:ring-primary-300"
      />
    </div>
  </div>
</DocumentFragment>
`;
